### ISSUE TITLE
Before step for plugins

### ISSUE THREAD

**Issue by siddharthist** [drone/drone#1443]
See discussion on #1441.

If a plugin could determine when drone moves on to the build step or does not, all kinds of security/access control features could easily be introduced while keeping Drone's core small. 

This kind of plugin would have to run before any other step, have access to all kinds of information (who created the commit, what changes are included, has .drone.yml changed, etc.) and have the ability to choose which stages of the drone pipeline to execute. 

Some common use cases would be to restrict builds to only certain users or users in a single github org, but these conditions and actions should be arbitrary to allow for greater flexibility.

**Comment by siddharthist** [comment]
This would even allow for a strictly smaller amount of code in the drone core, as the functionality involved in monitoring the checksum of `.drone.yml` with `drone.sec` could be exported into a plugin, as could the configuration of not building PRs on public repos. (In fact, these might be the best first example plugins to show off this functionality).

**Comment by bradrydzewski** [comment]
I'm thinking the easiest way to integrate this sort of change would be to have some sort of `before` section to the yaml that runs a set of plugins before the build starts:

```
before:
  check_trusted:
    when:
      event: pull_request
```

I'm not sure what an example plugin would look like, but it might be nice to kick around some ideas for a real-world "trusted author" plugin (related to security and trusted users) so that we can make sure the above approach works and there aren't any gaps.

**Comment by gtaylor** [comment]
We are starting to accumulate a healthy pile of sections, though. I wonder if there gets to be a point where we just figure out how to execute an arbitrary list of plugins in a certain order, or define dependencies and do away with the sections.

**Comment by bradrydzewski** [comment]
> This would even allow for a strictly smaller amount of code in the drone core, as the functionality involved in monitoring the checksum of .drone.yml with drone.sec could be exported into a plugin

@siddharthist the idea of moving the `.drone.sec` file to a plugin is appealing indeed. The only catch is that the `.drone.sec` file is parsed before the yaml is parsed, and the contents are injected into the yaml using simple string find / replace.

That being said, the `.drone.sec` contents could be unpacked and dumped to files that could be read by the other steps in the workflow, instead of using string interpolation in the yaml. I think this is something we need to explore further. Any ideas for improving the `.drone.sec` should probably be consolidated to https://github.com/drone/drone/issues/1250

> We are starting to accumulate a healthy pile of sections, though. I wonder if there gets to be a point where we just figure out how to execute an arbitrary list of plugins in a certain order, or define dependencies and do away with the sections.

YES I completely agree and this is something that needs to be addressed after the 0.5 release (pull queue implementation) in the 0.6 release. In fact, I think these are the last major change required before we can move from beta to release candidate.

**Comment by siddharthist** [comment]
@bradrydzewski What changes would be necessary to drone to make this feasible? Could you perhaps provide an outline/plan for implementation? I'd be happy to help, but I don't know where to start...

**Comment by donny-dont** [comment]
I still think this is going to make the YAML parsing nuts. @bradrydzewski I know we talked about this but I can't seem to find an associated issue so my suggestion is that Drone supports a imperative path as well as the current declarative path. This is similar to a grunt/gulp scenario from the Node world.

So as an example what if I had.

```
var environment = drone.getEnvironment();

var pipeline = new Pipeline();
pipeline.add(new BuildStep('unit tests', 'sh unit_tests.sh'));

if (environment.branch == 'release') {
  pipeline.add(new BuildStep('functional tests', 'sh functional_tests.sh');
}

var result = pipeline.run();

if (result == drone.SUCCESS) {
  var publishNpm = new PublishNpm(environment.NPM_REGISTRY, ...);

  if (publishNpm.run() == drone.SUCCESS) {
    var notify = new NotifySlack(environment.SLACK_TOKEN, ...);

    notify.run();
  }
}
```

**Comment by bradrydzewski** [comment]
@donny-dont I completely agree.

It has become apparent that the current yaml has some flaws that need to be addressed and I don't want to release and freeze a 1.0 until they are resolved. I think this will take time to prototype and see what works and what doesn't.

I think we need to give @siddharthist some options to integrate this new class of plugin today, without having to wait for a yaml or build configuration re-design. So even though I cringe at the idea of introducing a new section, which I actively want to remove, it might be the best way to move forward until we have a new design.

Speaking of new yaml design with scripting options, I think we need to come up with an RFC for these changes. One thing I want to avoid is getting 100 different opinions on minor variations of yaml formats. We need to approach this with a set of goals that any proposed format would need to meet.

I've been intrigued by some ideas presented by @cusspvz in https://github.com/drone/drone/issues/1356#issuecomment-170650041 that treated the build as a single set of steps. I started playing around with some variations such as the following:

```yaml
pipeline:
  clone:
    plugin: git

  cache:
    plugin: cache
    mount: [ node_modules ]

  mysql:
    image: mysql
    detach: true

  postgres:
    image: mysql
    detach: true

  build:
    image: golang
    commands:
      - go build
      - go get

  deploy:
    plugin: heroku
    app: foo
    token: $$HEROKU_TOKEN
    when:
      event: [ push, tag ]
      branch: master

catch:
  rollback:
    plugin: heroku_rollback
    token: $$HEROKU_TOKEN
    app: foo

finally:
  notify:
    plugin: slack
    token: $$SLACK_TOKEN
```

**Comment by donny-dont** [comment]
I like that it declares the pipeline. I'm just thinking things are going to get weird when you start throwing in when statements everywhere.

**Comment by bradrydzewski** [comment]
@donny-dont I think perhaps this is where a JS / scriptable option would be beneficial. Tools like gradle and gulp have show programming languages can work very well for build configuration.

**Comment by gtaylor** [comment]
Lua would be really sweet for this, but that's coming from a JS hater.

**Comment by donny-dont** [comment]
I think there's a Golang JavaScript interpreter so easy to roll that in

**Comment by donny-dont** [comment]
Appears theres a Lua implementation https://github.com/Shopify/go-lua as well as a JavaScript one https://github.com/robertkrimen/otto.

**Comment by siddharthist** [comment]
Hey all, it might be useful to make another issue for tracking ideas for improving the YAML spec. Seems like a lot of good ideas floating around!

**Comment by siddharthist** [comment]
Here's my outline of what needs to happen to add a homogenous plugin-based `before` step. Could somebody please review and let me know what's missing? Then I can get to work on implementing it.
```
Changes:
 1. Add another step to [Config][Config], which is `Before PluginSlice`
 2. Add another Node type, [`NodeBefore`][NodeBefore] that is created with
 `newPluginNode`.
 3. Add another rule to [Load][Load], specifically "Before." that looks a lot
 like "Plugin.", except is attached to `conf.Before.Slice()` and has type
 `NodeBefore`.

Tests:
 1. Add another test case with a "Before" clause to [parse\_test.go][parse_test]

[Config]: https://github.com/drone/drone-exec/blob/7dc4524fc4c8115d1dcc6cc5027f75606a3069ea/yaml/types.go#L5
[NodeBefore]: https://github.com/drone/drone-exec/blob/7dc4524fc4c8115d1dcc6cc5027f75606a3069ea/parser/node.go#L25
[Load]: https://github.com/drone/drone-exec/blob/7dc4524fc4c8115d1dcc6cc5027f75606a3069ea/parser/parse.go#L35

[parse_test]: https://github.com/drone/drone-exec/blob/7dc4524fc4c8115d1dcc6cc5027f75606a3069ea/yaml/parse_test.go
```

**Comment by cusspvz** [comment]
Wouldn't be significantly better to add a `catch` plugin that would catch errors from the latest `stage`?

The example I've provided on https://github.com/drone/drone/issues/1356#issuecomment-170650041 has already piping control, just because stages are processed on series and jobs on parallel.

I also don't think you need to include jobs into a key, such as `parallel` or `jobs`. If we set a list of names users can't use on jobs, such as: `stages`, `image` and some others; you can gather everything as jobs (plugins would be plugged on each job, script runner would be a plugin as well).

>  I think we need to come up with an RFC for these changes

I agree.

**Comment by bradrydzewski** [comment]
> Wouldn't be significantly better to add a catch plugin that would catch errors from the latest stage?

I'm not sure how this would work. This isn't really what plugins are used for. Have you really dug into the Drone plugin model and build runner? I ask because I respect your ideas (I put a lot of thought into your comment) but to move forward with an RFC Yaml improvements it will require a deep understanding of how the build runner and plugins work, and how Drone orchestrates containers.

**Comment by siddharthist** [comment]
I've just made a new issue for the purposes of discussing a new config format :)

**Comment by bradrydzewski** [comment]
@siddharthist your outline is pretty spot on. The only thing I would include is that we have a bunch of rules to transform the yaml before execution, setting default values, etc.
https://github.com/drone/drone-exec/blob/master/parser/funcs.go

**Comment by siddharthist** [comment]
@bradrydzewski Thanks for the review! Do you think that those would have to be altered to make this change?

**Comment by bradrydzewski** [comment]
@siddharthist I looked through the codebase quickly and don't think any rules will need to be adjusted. I can do a deeper review once we have a PR

**Comment by cusspvz** [comment]
> I'm not sure how this would work. Plugins are isolated Docker containers that run a task and return an exit code so handling orchestration is out of their scope / capability

Major problem I can see on current implementation is that stages are already pre-defined and all the code base relias on that definition, implementation of catchers with current Drone's logic would have to be written on the same way as well. I really would like to help you on that but unfortunately, inspite I can read and understand Go I cannot write awesomeness as I do with Javascript.

> Have you had a chance to really dig into the Drone plugin model and build runner?

Yes I did, in fact, I had took almost an hour reading the logic of the most important repos (such as cli, job's logic and so on).

> I hope this doesn't come off as negative (not my intention). I think you have good ideas and I want you to really understand the guts of Drone and how it works, and how it is different than GitLab, since I think it will really help us come up with a better proposal

Don't worry with that, I like to keep things positive and to understand others opinions. :) You already said this sort of major changes won't be able to be released until *2.0*, and I completely understand your point on keeping things stable until *1.0* and yet your concern to enhance current yaml design.

Maybe we can have two separate discussion spots for describing RFC yaml drafts for *0.x -> 1.0* and *2.0* versions.
